# Deathstar.py
Deathstar.py is a tool currently in development that can be used specifically to analyze the outputs from the tool CNVkit, this tool takes the cnr files, preferentially from a cohort and processess them together to generate plots for easier visualization of alterations in the genome, one has to put their gene list as in an array and feed it to the loop, along with the cnr files, most of the other tasks are automated, except in some places where manual ovveride is needed to filter amplifications and deletions based on the z score methods and to change the thresholds as needed by the user. 
# How does Deathstar works?
It takes in the cnr files and the gene names as an input and starts by plotting a line plot for each and every bin across each sample and a median plot for all the samples for better visualization. It also has a functionality which solves a major problem in WES, probes which cover an area which do not lie over an exon, were removed by checking for overlaps of the probe and exon area for each and every gene. Currently it only considers those probes which have a 100% overlap with an exon, we will have to check how to approach if there is a 50%-90% overlap. Then it plots for depth and standard deviation which tells us more about how low or high is the variance among the samples for that particular bin (Bins are nothing but genomic regions covered by the probes or the exonic regions captured during sequencing) and is easier to interpret which bin is much more trustable and vice versa. Then it uses Mann-Whitney test to check for significant differences in the between each and every bin and trying to identify which bin stands out from all the other bins. Then it has a z score filtering functionality which helps us to filter out bins which dont follow the normal pattern and are a few standard deviations away from the sample wide mean. We have set a threshold of 2.5 for the z score meaning we are going to consider the bins of those genes who's z score is more than and less than +2.5 and -2.5. Meaning these points are 2.5 standard deviations away from the sample wide mean. These regions can then be further inspected in IGV to check for abnormalities in the coverage depth. This code might need additional packages such as num2words.
```bash
pip install num2words
```
# Metagenomics.sh
Metagenomics.sh is a metagenomics pipeline made out of qiime2 which has a lot of cutomizations that can be enabled and a lot of tools to choose from, like the denoisers or phylogenetic tree creation. This pipeline is still under development, with further improvements being done to provide customized plots based out of R and more downstream analysis for better understanding of the data.
# test.nf and wes.sh
These are the same pipelines in two different implementation for the WES data processing, made according to the GATK best practices and some more additonal analysis preseent in the bash file.
